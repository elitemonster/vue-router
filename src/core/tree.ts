import type { ResolvedOptions } from '../options'
import { createTreeLeafValue, TreeRouteParam } from './treeLeafValue'
import type { TreeLeafValue } from './treeLeafValue'
import { trimExtension } from './utils'
import { CustomRouteBlock } from './customBlock'

export class TreeLeaf {
  /**
   * value of the node
   */
  value: TreeLeafValue

  /**
   * children of the node
   */
  children: Map<string, TreeLeaf> = new Map()

  /**
   * Parent node.
   */
  parent?: TreeLeaf

  /**
   * Plugin options taken into account by the tree.
   */
  options: ResolvedOptions

  constructor(options: ResolvedOptions, filePath: string, parent?: TreeLeaf) {
    this.options = options
    this.parent = parent
    this.value = createTreeLeafValue(filePath, parent?.value)
  }

  /**
   * Adds a path to the tree. `path` cannot start with a `/`.
   *
   * @param path - route path of the file
   * @param filePath - file path, defaults to path for convenience and testing
   */
  insert(path: string, filePath: string = path): TreeLeaf {
    const { tail, segment, viewName, isComponent } = splitFilePath(path)

    if (!this.children.has(segment)) {
      this.children.set(segment, new TreeLeaf(this.options, segment, this))
    }
    const child = this.children.get(segment)!

    if (isComponent) {
      child.value.filePaths.set(viewName, filePath)
    }

    if (tail) {
      return child.insert(tail, filePath)
    }
    return child
  }

  setCustomRouteBlock(path: string, routeBlock: CustomRouteBlock | undefined) {
    this.value.setOverride(path, routeBlock)
  }

  getSortedChildren() {
    return Array.from(this.children.values()).sort((a, b) =>
      a.path.localeCompare(b.path)
    )
  }

  remove(path: string) {
    const { tail, segment, viewName, isComponent } = splitFilePath(path)

    const child = this.children.get(segment)
    if (!child) {
      throw new Error(
        `Cannot Delete "${path}". "${segment}" not found at "${this.path}".`
      )
    }

    if (tail) {
      child.remove(tail)
      // if the child doesn't create any route
      if (child.children.size === 0 && child.value.filePaths.size === 0) {
        this.children.delete(segment)
      }
    } else {
      // it can only be component because we only listen for removed files, not folders
      if (isComponent) {
        child.value.filePaths.delete(viewName)
      }
      // this is the file we wanted to remove
      if (child.children.size === 0 && child.value.filePaths.size === 0) {
        this.children.delete(segment)
      }
    }
  }

  get path() {
    return (
      this.value.overrides.path ??
      (this.parent?.isRoot() ? '/' : '') + this.value.pathSegment
    )
  }

  get fullPath() {
    return this.value.overrides.path ?? this.value.path
  }

  get name() {
    return this.value.overrides.name || this.options.getRouteName(this)
  }

  get meta() {
    const overrideMeta = { ...this.value.overrides.meta }

    if (this.value.includeLoaderGuard) {
      overrideMeta._loaderGuard = true
    }

    return Object.keys(overrideMeta).length > 0
      ? JSON.stringify(overrideMeta, null, 2)
      : ''
  }

  get params(): TreeRouteParam[] {
    const params = this.value.isParam() ? [...this.value.params] : []
    let node = this.parent
    while (node) {
      if (node.value.isParam()) {
        params.unshift(...node.value.params)
      }
      node = node.parent
    }

    return params
  }

  isRoot() {
    return this.value.path === '/' && !this.value.filePaths.size
  }

  toString(): string {
    return `${this.value}${
      this.value.filePaths.size
        ? ` ðŸ“„ (${Array.from(this.value.filePaths.keys()).join(', ')})`
        : ''
    }`
  }
}

export function createPrefixTree(options: ResolvedOptions) {
  return new TreeLeaf(options, '')
}

/**
 * Splits a path into by finding the first '/' and returns the tail and segment. If it has an extension, it removes it.
 * If it contains a named view, it returns the view name as well (otherwise it's default).
 *
 * @param filePath - filePath to split
 */
function splitFilePath(filePath: string) {
  const slashPos = filePath.indexOf('/')
  let head = slashPos < 0 ? filePath : filePath.slice(0, slashPos)
  const tail = slashPos < 0 ? '' : filePath.slice(slashPos + 1)

  let segment = head
  // only the last segment can be a filename with an extension
  if (!tail) {
    segment = trimExtension(head)
  }
  let viewName = 'default'

  const namedSeparatorPos = segment.indexOf('@')

  if (namedSeparatorPos > 0) {
    viewName = segment.slice(namedSeparatorPos + 1)
    segment = segment.slice(0, namedSeparatorPos)
  }

  const isComponent = segment !== head

  return {
    segment,
    tail,
    viewName,
    isComponent,
  }
}
