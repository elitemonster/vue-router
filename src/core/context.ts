import chokidar from 'chokidar'
import { resolve } from 'path'
import { Options } from '../types'
import { createPrefixTree } from './tree'
import { promises as fs } from 'fs'
import { logTree, throttle } from './utils'
import { generateRouteNamedMap } from './generateRouteMap'

export function createRoutesContext(options: Required<Options>) {
  const { dts: preferDTS, root } = options
  const dts =
    preferDTS === false
      ? false
      : preferDTS === true
      ? resolve(root, 'typed-router.d.ts')
      : resolve(root, preferDTS)

  const routeTree = createPrefixTree()

  const resolvedRoutesFolder = resolve(root, options.routesFolder)
  const serverWatcher = chokidar.watch(resolvedRoutesFolder, {
    // TODO: create a scanRouteFolders() function that also works for build
    // ignoreInitial: true,
    disableGlobbing: true,
    ignorePermissionErrors: true,
    // useFsEvents: true,
    // TODO: allow user options
  })

  function stripRouteFolder(path: string) {
    return path.slice(resolvedRoutesFolder.length + 1)
  }

  function setupWatcher() {
    serverWatcher
      .on('change', (path) => {
        // TODO: parse defineRoute macro?
        console.log('change', path)
        writeConfigFiles()
      })
      .on('add', (path) => {
        console.log('added', path)
        routeTree.insert(
          stripRouteFolder(path),
          // './' + path
          resolve(root, path)
        )
        writeConfigFiles()
      })
      .on('unlink', (path) => {
        console.log('remove', path)
        routeTree.remove(stripRouteFolder(path))
        writeConfigFiles()
      })
  }

  function stop() {
    serverWatcher.close()
  }

  function generateRoutes() {
    return `export const routes = ${routeTree.toRouteRecordString()}`
  }

  function generateDTS(): string {
    return `
// Generated by unplugin-vue-router. ‼️ DO NOT MODIFY THIS FILE ‼️
// It's recommended to commit this file.
// Make sure to add this file to your tsconfig.json file as an "includes" or "files" entry.

/// <reference types="unplugin-vue-router/client" />

import type {
  RouteRecordInfo,
  _ParamValue,
  _ParamValueOneOrMore,
  _ParamValueZeroOrMore,
  _ParamValueZeroOrOne,
} from 'unplugin-vue-router'

declare module '~routes' {
${generateRouteNamedMap(routeTree)
  .split('\n')
  .filter((line) => line)
  .map((line) => line.padStart(2))
  .join('\n')}
}

`
  }

  let lastDTS: string | undefined
  async function _writeConfigFiles() {
    console.log('writing')
    logTree(routeTree)
    if (dts) {
      const content = generateDTS()
      if (lastDTS !== content) {
        await fs.writeFile(dts, content, 'utf-8')
        lastDTS = content
      }
    }
  }
  const writeConfigFiles = throttle(_writeConfigFiles, 500)

  setupWatcher()

  return {
    stop,
    writeConfigFiles,

    generateRoutes,
  }
}
