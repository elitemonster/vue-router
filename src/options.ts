import { isPackageExists } from 'local-pkg'
import { getFileBasedRouteName, isArray } from './core/utils'
import type { TreeLeaf } from './core/tree'

export interface RoutesFolderOption {
  src: string
  path?: string
}

export type _RoutesFolder = string | RoutesFolderOption
export type RoutesFolder = _RoutesFolder[] | _RoutesFolder

export interface ResolvedOptions {
  /**
   * Extensions of files to be considered as pages. Defaults to `['.vue']`. Cannot be empty.
   */
  extensions: string[]

  /**
   * Folder containing the components that should be used for routes. Can also be an array if you want to add multiple
   * folders, or an object if you want to define a route prefix. Supports glob patterns but must be a folder, use
   * `extensions` and `exclude` to filter files.
   *
   * @default "src/pages"
   */
  routesFolder: RoutesFolder

  /**
   * Method to generate the name of a route.
   */
  getRouteName: (node: TreeLeaf) => string

  /**
   * EXPERIMENTAL: add the data fetching meta properties to generated routes.
   */
  dataFetching: boolean

  /**
   * Defines how page components should be imported. Defaults to dynamic imports to enable lazy loading of pages.
   */
  importMode: _OptionsImportMode

  /**
   * Array of file globs to ignore. Defaults to `[]`.
   */
  exclude: string[]

  /**
   * Root of the project. All paths are resolved relatively to this one. Defaults to `process.cwd()`.
   */
  root: string

  /**
   * Language for `<route>` blocks in SFC files. Defaults to `'json5'`.
   */
  routeBlockLang: string

  /**
   * Should generate d.ts files. Defaults to `true` if `typescript` is installed.
   */
  dts: boolean | string

  /**
   * Allows inspection by vite-plugin-inspect by not adding the leading `\0` to the id of virtual modules.
   * @internal
   */
  _inspect: boolean

  /**
   * Activates debug logs.
   */
  logs: boolean
}

/**
 * @internal
 */
export type _OptionsImportMode =
  | 'sync'
  | 'async'
  | ((filepath: string) => 'sync' | 'async')

export type Options = Partial<ResolvedOptions>

export const DEFAULT_OPTIONS: ResolvedOptions = {
  extensions: ['.vue'],
  exclude: [],
  routesFolder: 'src/pages',
  routeBlockLang: 'json5',
  getRouteName: getFileBasedRouteName,
  dataFetching: false,
  importMode: 'async',
  root: process.cwd(),
  dts: isPackageExists('typescript'),
  logs: false,
  _inspect: false,
}

export interface ServerContext {
  invalidate: (module: string) => void
  reload: () => void
}

export function normalizeRoutesFolderOption(
  routesFolder: ResolvedOptions['routesFolder']
): RoutesFolderOption[] {
  return (isArray(routesFolder) ? routesFolder : [routesFolder]).map(
    (routeOption) =>
      typeof routeOption === 'string' ? { src: routeOption } : routeOption
  )
}
