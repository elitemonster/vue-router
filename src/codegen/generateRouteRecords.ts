import type { TreeLeaf } from '../core/tree'
import { ResolvedOptions, _OptionsImportMode } from '../options'
import { basename } from 'pathe'

export function generateRouteRecord(
  node: TreeLeaf,
  options: ResolvedOptions,
  importList: Map<string, string>,
  indent = 0
): string {
  // root
  if (node.value.path === '/' && indent === 0) {
    return `[
${node
  .getSortedChildren()
  .map((child) => generateRouteRecord(child, options, importList, indent + 1))
  .join(',\n')}
]`
  }

  const startIndent = ' '.repeat(indent * 2)
  const indentStr = ' '.repeat((indent + 1) * 2)

  // TODO: should meta be defined a different way to allow preserving imports?
  // const meta = node.value.overrides.meta

  // path
  return `${startIndent}{
${indentStr}path: '${node.path}',
${indentStr}${
    node.value.filePaths.size ? `name: '${node.name}',` : '/* no name */'
  }
${
  // component
  indentStr
}${
    node.value.filePaths.size
      ? generateRouteRecordComponent(
          node,
          indentStr,
          options.importMode,
          importList
        )
      : '/* no component */'
  }
${
  // props
  indentStr
}${
    node.value.overrides.props != null
      ? `props: ${node.value.overrides.props},`
      : '/* no props */'
  }
${
  // children
  indentStr
}${
    node.children.size > 0
      ? `children: [
${node
  .getSortedChildren()
  .map((child) => generateRouteRecord(child, options, importList, indent + 2))
  .join(',\n')}
${indentStr}],`
      : '/* no children */'
  }${formatMeta(node, indentStr)}
${startIndent}}`
}

function generateRouteRecordComponent(
  node: TreeLeaf,
  indentStr: string,
  importMode: _OptionsImportMode,
  importList: Map<string, string>
): string {
  const files = Array.from(node.value.filePaths)
  const isDefaultExport = files.length === 1 && files[0][0] === 'default'
  return isDefaultExport
    ? `component: ${generatePageImport(files[0][1], importMode, importList)},`
    : // files has at least one entry
      `components: {
${files
  .map(
    ([key, path]) =>
      `${indentStr + '  '}'${key}': ${generatePageImport(
        path,
        importMode,
        importList
      )}`
  )
  .join(',\n')}
${indentStr}},`
}

/**
 * Generate the import (dynamic or static) for the given filepath. If the filepath is a static import, add it to the
 * @param filepath - the filepath to the file
 * @param importMode - the import mode to use
 * @param importList - the import list to fill
 * @returns
 */
function generatePageImport(
  filepath: string,
  importMode: _OptionsImportMode,
  importList: Map<string, string>
) {
  const mode =
    typeof importMode === 'function' ? importMode(filepath) : importMode
  if (mode === 'async') {
    return `() => import('${filepath}')`
  } else {
    const importName = `_page_${filepath.replace(/[\/\.]/g, '_')}`
    importList.set(filepath, importName)
    return importName
  }
}

function generateImportList(node: TreeLeaf, indentStr: string) {
  const files = Array.from(node.value.filePaths)

  return `[
${files
  .map(([_key, path]) => `${indentStr}  () => import('${path}')`)
  .join(',\n')}
${indentStr}]`
}

const LOADER_GUARD_RE = /['"]_loaderGuard['"]:.*$/

function formatMeta(node: TreeLeaf, indent: string): string {
  const meta = node.meta
  const formatted =
    meta &&
    meta
      .split('\n')
      .map(
        (line) =>
          indent +
          line.replace(
            LOADER_GUARD_RE,
            '[_LoaderSymbol]: ' + generateImportList(node, indent + '  ') + ','
          )
      )
      .join('\n')

  return formatted ? '\n' + indent + 'meta: ' + formatted.trimStart() : ''
}
