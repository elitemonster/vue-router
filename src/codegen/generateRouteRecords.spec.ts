import { basename } from 'pathe'
import { describe, expect, it } from 'vitest'
import { createPrefixTree, TreeNode } from '../core/tree'
import { ResolvedOptions, resolveOptions } from '../options'
import { generateRouteRecord } from './generateRouteRecords'

const DEFAULT_OPTIONS = resolveOptions({})

describe('generateRouteRecord', () => {
  function generateRouteRecordSimple(tree: TreeNode) {
    return generateRouteRecord(tree, DEFAULT_OPTIONS, new Map())
  }

  it('works with an empty tree', () => {
    const tree = createPrefixTree(DEFAULT_OPTIONS)

    expect(generateRouteRecordSimple(tree)).toMatchInlineSnapshot(`
      "[

      ]"
    `)
  })

  it('works with some paths at root', () => {
    const tree = createPrefixTree(DEFAULT_OPTIONS)
    tree.insert('a.vue')
    tree.insert('b.vue')
    tree.insert('c.vue')
    expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
  })

  it('handles multiple named views', () => {
    const tree = createPrefixTree(DEFAULT_OPTIONS)
    tree.insert('foo.vue')
    tree.insert('foo@a.vue')
    tree.insert('foo@b.vue')
    expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
  })

  it('handles single named views', () => {
    const tree = createPrefixTree(DEFAULT_OPTIONS)
    tree.insert('foo@a.vue')
    expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
  })

  it('nested children', () => {
    const tree = createPrefixTree(DEFAULT_OPTIONS)
    tree.insert('a/a.vue')
    tree.insert('a/b.vue')
    tree.insert('a/c.vue')
    tree.insert('b/b.vue')
    tree.insert('b/c.vue')
    tree.insert('b/d.vue')
    expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
    tree.insert('c.vue')
    tree.insert('d.vue')
    expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
  })

  it('adds children and name when folder and component exist', () => {
    const tree = createPrefixTree(DEFAULT_OPTIONS)
    tree.insert('a/c.vue')
    tree.insert('b/c.vue')
    tree.insert('a.vue')
    tree.insert('d.vue')
    expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
  })

  it('correctly names index.vue files', () => {
    const tree = createPrefixTree(DEFAULT_OPTIONS)
    tree.insert('index.vue')
    tree.insert('b/index.vue')
    expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
  })

  it('handles non nested routes', () => {
    const tree = createPrefixTree(DEFAULT_OPTIONS)
    tree.insert('users.vue')
    tree.insert('users/index.vue')
    tree.insert('users/other.vue')
    tree.insert('users.not-nested.vue')
    tree.insert('users/[id]/index.vue')
    tree.insert('users/[id]/other.vue')
    tree.insert('users/[id].vue')
    tree.insert('users/[id].not-nested.vue')
    tree.insert('users.[id].also-not-nested.vue')
    expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
  })

  it('generate static imports', () => {
    const options: ResolvedOptions = {
      ...DEFAULT_OPTIONS,
      importMode: 'sync',
    } as const
    const tree = createPrefixTree(options)
    tree.insert('a.vue')
    tree.insert('b.vue')
    tree.insert('nested/file/c.vue')
    const importList = new Map<string, string>()
    expect(generateRouteRecord(tree, options, importList)).toMatchSnapshot()

    expect(importList).toMatchSnapshot()
  })

  it('generate custom imports', () => {
    const options: ResolvedOptions = {
      ...DEFAULT_OPTIONS,
      importMode: (filepath) =>
        basename(filepath) === 'a.vue' ? 'sync' : 'async',
    }

    const tree = createPrefixTree(options)
    tree.insert('a.vue')
    tree.insert('b.vue')
    tree.insert('nested/file/c.vue')
    const importList = new Map<string, string>()
    expect(generateRouteRecord(tree, options, importList)).toMatchSnapshot()

    expect(importList).toMatchSnapshot()
  })

  describe('names', () => {
    it('creates single word names', () => {
      const tree = createPrefixTree(DEFAULT_OPTIONS)
      tree.insert('index.vue')
      tree.insert('about.vue')
      tree.insert('users/index.vue')
      tree.insert('users/[id].vue')
      tree.insert('users/[id]/edit.vue')
      tree.insert('users/new.vue')

      expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
    })

    it('creates multi word names', () => {
      const tree = createPrefixTree(DEFAULT_OPTIONS)
      tree.insert('index.vue')
      tree.insert('my-users.vue')
      tree.insert('MyPascalCaseUsers.vue')
      tree.insert('some-nested/file-with-[id]-in-the-middle.vue')

      expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
    })

    it('works with nested views', () => {
      const tree = createPrefixTree(DEFAULT_OPTIONS)
      tree.insert('index.vue')
      tree.insert('users.vue')
      tree.insert('users/index.vue')
      tree.insert('users/[id]/edit.vue')
      tree.insert('users/[id].vue')

      expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
    })
  })

  describe('route block', () => {
    it('adds meta data', async () => {
      const tree = createPrefixTree(DEFAULT_OPTIONS)
      const node = tree.insert('index.vue')
      node.setCustomRouteBlock('index.vue', {
        meta: {
          auth: true,
          title: 'Home',
        },
      })

      expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
    })

    it('merges multiple meta properties', async () => {
      const tree = createPrefixTree(DEFAULT_OPTIONS)
      const node = tree.insert('index.vue')
      node.setCustomRouteBlock('index.vue', {
        path: '/custom',
        meta: {
          one: true,
        },
      })
      node.setCustomRouteBlock('index@named.vue', {
        name: 'hello',
        meta: {
          two: true,
        },
      })

      expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
    })

    it('merges regardless of order', async () => {
      const tree = createPrefixTree(DEFAULT_OPTIONS)
      const node = tree.insert('index.vue')
      node.setCustomRouteBlock('index.vue', {
        name: 'a',
      })
      node.setCustomRouteBlock('index@named.vue', {
        name: 'b',
      })

      const one = generateRouteRecordSimple(tree)

      node.setCustomRouteBlock('index@named.vue', {
        name: 'b',
      })
      node.setCustomRouteBlock('index.vue', {
        name: 'a',
      })

      expect(generateRouteRecordSimple(tree)).toBe(one)

      expect(one).toMatchSnapshot()
    })

    it('handles named views with empty route blocks', () => {
      const tree = createPrefixTree(DEFAULT_OPTIONS)
      const node = tree.insert('index.vue')
      const n2 = tree.insert('index@named.vue')
      expect(node).toBe(n2)
      // coming from index.vue
      node.setCustomRouteBlock('index.vue', {
        meta: {
          auth: true,
          title: 'Home',
        },
      })
      // coming from index@named.vue (no route block)
      node.setCustomRouteBlock('index@named.vue', undefined)

      expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
    })

    // FIXME: allow aliases
    it.todo('merges alias properties', async () => {
      const tree = createPrefixTree(DEFAULT_OPTIONS)
      const node = tree.insert('index.vue')
      node.setCustomRouteBlock('index.vue', {
        alias: '/one',
      })
      node.setCustomRouteBlock('index@named.vue', {
        alias: ['/two', '/three'],
      })

      expect(generateRouteRecordSimple(tree)).toMatchInlineSnapshot(`
        "[
          {
            path: '/',
            name: '/',
            component: () => import('index.vue'),
            /* no props */
            /* no children */
          }
        ]"
      `)
    })

    it('merges deep meta properties', async () => {
      const tree = createPrefixTree(DEFAULT_OPTIONS)
      const node = tree.insert('index.vue')
      node.setCustomRouteBlock('index.vue', {
        meta: {
          a: { one: 1 },
          b: { a: [2] },
        },
      })
      node.setCustomRouteBlock('index@named.vue', {
        meta: {
          a: { two: 1 },
          b: { a: [3] },
        },
      })

      expect(generateRouteRecordSimple(tree)).toMatchSnapshot()
    })
  })
})
