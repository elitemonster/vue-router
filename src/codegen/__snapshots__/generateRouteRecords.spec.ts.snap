// Vitest Snapshot v1

exports[`generateRouteRecord > adds children and name when folder and component exist 1`] = `
"[
  {
    path: '/a',
    name: '/a',
    component: () => import('a.vue'),
    /* no props */
    children: [
      {
        path: 'c',
        name: '/a/c',
        component: () => import('a/c.vue'),
        /* no props */
        /* no children */
      }
    ],
  },
  {
    path: '/b',
    /* no name */
    /* no component */
    /* no props */
    children: [
      {
        path: 'c',
        name: '/b/c',
        component: () => import('b/c.vue'),
        /* no props */
        /* no children */
      }
    ],
  },
  {
    path: '/d',
    name: '/d',
    component: () => import('d.vue'),
    /* no props */
    /* no children */
  }
]"
`;

exports[`generateRouteRecord > correctly names index.vue files 1`] = `
"[
  {
    path: '/',
    name: '/',
    component: () => import('index.vue'),
    /* no props */
    /* no children */
  },
  {
    path: '/b',
    /* no name */
    /* no component */
    /* no props */
    children: [
      {
        path: '',
        name: '/b/',
        component: () => import('b/index.vue'),
        /* no props */
        /* no children */
      }
    ],
  }
]"
`;

exports[`generateRouteRecord > generate custom imports 1`] = `
"[
  {
    path: '/a',
    name: '/a',
    component: _page_a_vue,
    /* no props */
    /* no children */
  },
  {
    path: '/b',
    name: '/b',
    component: () => import('b.vue'),
    /* no props */
    /* no children */
  },
  {
    path: '/nested',
    /* no name */
    /* no component */
    /* no props */
    children: [
      {
        path: 'file',
        /* no name */
        /* no component */
        /* no props */
        children: [
          {
            path: 'c',
            name: '/nested/file/c',
            component: () => import('nested/file/c.vue'),
            /* no props */
            /* no children */
          }
        ],
      }
    ],
  }
]"
`;

exports[`generateRouteRecord > generate custom imports 2`] = `
Map {
  "a.vue" => "_page_a_vue",
}
`;

exports[`generateRouteRecord > generate static imports 1`] = `
"[
  {
    path: '/a',
    name: '/a',
    component: _page_a_vue,
    /* no props */
    /* no children */
  },
  {
    path: '/b',
    name: '/b',
    component: _page_b_vue,
    /* no props */
    /* no children */
  },
  {
    path: '/nested',
    /* no name */
    /* no component */
    /* no props */
    children: [
      {
        path: 'file',
        /* no name */
        /* no component */
        /* no props */
        children: [
          {
            path: 'c',
            name: '/nested/file/c',
            component: _page_nested_file_c_vue,
            /* no props */
            /* no children */
          }
        ],
      }
    ],
  }
]"
`;

exports[`generateRouteRecord > generate static imports 2`] = `
Map {
  "a.vue" => "_page_a_vue",
  "b.vue" => "_page_b_vue",
  "nested/file/c.vue" => "_page_nested_file_c_vue",
}
`;

exports[`generateRouteRecord > handles multiple named views 1`] = `
"[
  {
    path: '/foo',
    name: '/foo',
    components: {
      'default': () => import('foo.vue'),
      'a': () => import('foo@a.vue'),
      'b': () => import('foo@b.vue')
    },
    /* no props */
    /* no children */
  }
]"
`;

exports[`generateRouteRecord > handles non nested routes 1`] = `
"[
  {
    path: '/users',
    name: '/users',
    component: () => import('users.vue'),
    /* no props */
    children: [
      {
        path: '',
        name: '/users/',
        component: () => import('users/index.vue'),
        /* no props */
        /* no children */
      },
      {
        path: ':id',
        name: '/users/[id]',
        component: () => import('users/[id].vue'),
        /* no props */
        children: [
          {
            path: '',
            name: '/users/[id]/',
            component: () => import('users/[id]/index.vue'),
            /* no props */
            /* no children */
          },
          {
            path: 'other',
            name: '/users/[id]/other',
            component: () => import('users/[id]/other.vue'),
            /* no props */
            /* no children */
          }
        ],
      },
      {
        path: ':id/not-nested',
        name: '/users/[id].not-nested',
        component: () => import('users/[id].not-nested.vue'),
        /* no props */
        /* no children */
      },
      {
        path: 'other',
        name: '/users/other',
        component: () => import('users/other.vue'),
        /* no props */
        /* no children */
      }
    ],
  },
  {
    path: '/users/:id/also-not-nested',
    name: '/users.[id].also-not-nested',
    component: () => import('users.[id].also-not-nested.vue'),
    /* no props */
    /* no children */
  },
  {
    path: '/users/not-nested',
    name: '/users.not-nested',
    component: () => import('users.not-nested.vue'),
    /* no props */
    /* no children */
  }
]"
`;

exports[`generateRouteRecord > handles single named views 1`] = `
"[
  {
    path: '/foo',
    name: '/foo',
    components: {
      'a': () => import('foo@a.vue')
    },
    /* no props */
    /* no children */
  }
]"
`;

exports[`generateRouteRecord > names > creates multi word names 1`] = `
"[
  {
    path: '/',
    name: '/',
    component: () => import('index.vue'),
    /* no props */
    /* no children */
  },
  {
    path: '/my-users',
    name: '/my-users',
    component: () => import('my-users.vue'),
    /* no props */
    /* no children */
  },
  {
    path: '/MyPascalCaseUsers',
    name: '/MyPascalCaseUsers',
    component: () => import('MyPascalCaseUsers.vue'),
    /* no props */
    /* no children */
  },
  {
    path: '/some-nested',
    /* no name */
    /* no component */
    /* no props */
    children: [
      {
        path: 'file-with-:id-in-the-middle',
        name: '/some-nested/file-with-[id]-in-the-middle',
        component: () => import('some-nested/file-with-[id]-in-the-middle.vue'),
        /* no props */
        /* no children */
      }
    ],
  }
]"
`;

exports[`generateRouteRecord > names > creates single word names 1`] = `
"[
  {
    path: '/',
    name: '/',
    component: () => import('index.vue'),
    /* no props */
    /* no children */
  },
  {
    path: '/about',
    name: '/about',
    component: () => import('about.vue'),
    /* no props */
    /* no children */
  },
  {
    path: '/users',
    /* no name */
    /* no component */
    /* no props */
    children: [
      {
        path: '',
        name: '/users/',
        component: () => import('users/index.vue'),
        /* no props */
        /* no children */
      },
      {
        path: ':id',
        name: '/users/[id]',
        component: () => import('users/[id].vue'),
        /* no props */
        children: [
          {
            path: 'edit',
            name: '/users/[id]/edit',
            component: () => import('users/[id]/edit.vue'),
            /* no props */
            /* no children */
          }
        ],
      },
      {
        path: 'new',
        name: '/users/new',
        component: () => import('users/new.vue'),
        /* no props */
        /* no children */
      }
    ],
  }
]"
`;

exports[`generateRouteRecord > names > works with nested views 1`] = `
"[
  {
    path: '/',
    name: '/',
    component: () => import('index.vue'),
    /* no props */
    /* no children */
  },
  {
    path: '/users',
    name: '/users',
    component: () => import('users.vue'),
    /* no props */
    children: [
      {
        path: '',
        name: '/users/',
        component: () => import('users/index.vue'),
        /* no props */
        /* no children */
      },
      {
        path: ':id',
        name: '/users/[id]',
        component: () => import('users/[id].vue'),
        /* no props */
        children: [
          {
            path: 'edit',
            name: '/users/[id]/edit',
            component: () => import('users/[id]/edit.vue'),
            /* no props */
            /* no children */
          }
        ],
      }
    ],
  }
]"
`;

exports[`generateRouteRecord > nested children 1`] = `
"[
  {
    path: '/a',
    /* no name */
    /* no component */
    /* no props */
    children: [
      {
        path: 'a',
        name: '/a/a',
        component: () => import('a/a.vue'),
        /* no props */
        /* no children */
      },
      {
        path: 'b',
        name: '/a/b',
        component: () => import('a/b.vue'),
        /* no props */
        /* no children */
      },
      {
        path: 'c',
        name: '/a/c',
        component: () => import('a/c.vue'),
        /* no props */
        /* no children */
      }
    ],
  },
  {
    path: '/b',
    /* no name */
    /* no component */
    /* no props */
    children: [
      {
        path: 'b',
        name: '/b/b',
        component: () => import('b/b.vue'),
        /* no props */
        /* no children */
      },
      {
        path: 'c',
        name: '/b/c',
        component: () => import('b/c.vue'),
        /* no props */
        /* no children */
      },
      {
        path: 'd',
        name: '/b/d',
        component: () => import('b/d.vue'),
        /* no props */
        /* no children */
      }
    ],
  }
]"
`;

exports[`generateRouteRecord > nested children 2`] = `
"[
  {
    path: '/a',
    /* no name */
    /* no component */
    /* no props */
    children: [
      {
        path: 'a',
        name: '/a/a',
        component: () => import('a/a.vue'),
        /* no props */
        /* no children */
      },
      {
        path: 'b',
        name: '/a/b',
        component: () => import('a/b.vue'),
        /* no props */
        /* no children */
      },
      {
        path: 'c',
        name: '/a/c',
        component: () => import('a/c.vue'),
        /* no props */
        /* no children */
      }
    ],
  },
  {
    path: '/b',
    /* no name */
    /* no component */
    /* no props */
    children: [
      {
        path: 'b',
        name: '/b/b',
        component: () => import('b/b.vue'),
        /* no props */
        /* no children */
      },
      {
        path: 'c',
        name: '/b/c',
        component: () => import('b/c.vue'),
        /* no props */
        /* no children */
      },
      {
        path: 'd',
        name: '/b/d',
        component: () => import('b/d.vue'),
        /* no props */
        /* no children */
      }
    ],
  },
  {
    path: '/c',
    name: '/c',
    component: () => import('c.vue'),
    /* no props */
    /* no children */
  },
  {
    path: '/d',
    name: '/d',
    component: () => import('d.vue'),
    /* no props */
    /* no children */
  }
]"
`;

exports[`generateRouteRecord > route block > adds meta data 1`] = `
"[
  {
    path: '/',
    name: '/',
    component: () => import('index.vue'),
    /* no props */
    /* no children */
    meta: {
      \\"auth\\": true,
      \\"title\\": \\"Home\\"
    }
  }
]"
`;

exports[`generateRouteRecord > route block > handles named views with empty route blocks 1`] = `
"[
  {
    path: '/',
    name: '/',
    components: {
      'default': () => import('index.vue'),
      'named': () => import('index@named.vue')
    },
    /* no props */
    /* no children */
    meta: {
      \\"auth\\": true,
      \\"title\\": \\"Home\\"
    }
  }
]"
`;

exports[`generateRouteRecord > route block > merges deep meta properties 1`] = `
"[
  {
    path: '/',
    name: '/',
    component: () => import('index.vue'),
    /* no props */
    /* no children */
    meta: {
      \\"a\\": {
        \\"one\\": 1,
        \\"two\\": 1
      },
      \\"b\\": {
        \\"a\\": [
          2,
          3
        ]
      }
    }
  }
]"
`;

exports[`generateRouteRecord > route block > merges multiple meta properties 1`] = `
"[
  {
    path: '/custom',
    name: 'hello',
    component: () => import('index.vue'),
    /* no props */
    /* no children */
    meta: {
      \\"one\\": true,
      \\"two\\": true
    }
  }
]"
`;

exports[`generateRouteRecord > route block > merges regardless of order 1`] = `
"[
  {
    path: '/',
    name: 'b',
    component: () => import('index.vue'),
    /* no props */
    /* no children */
  }
]"
`;

exports[`generateRouteRecord > works with some paths at root 1`] = `
"[
  {
    path: '/a',
    name: '/a',
    component: () => import('a.vue'),
    /* no props */
    /* no children */
  },
  {
    path: '/b',
    name: '/b',
    component: () => import('b.vue'),
    /* no props */
    /* no children */
  },
  {
    path: '/c',
    name: '/c',
    component: () => import('c.vue'),
    /* no props */
    /* no children */
  }
]"
`;
